ConfigurationManager no longer throws if _configRetriever fails but _currentConfiguration is not null

mafurman opened the pull request:
Fixes #1111 and #1112.

brentschmaltz commented on src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager.cs:
```diff
@@ -199,7 +199,10 @@ public async Task<T> GetConfigurationAsync(CancellationToken cancel)
+                        if (_currentConfiguration == null) // Throw an exception if there's no configuration to return.
+                            throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX20803, (_metadataAddress ?? "null")), ex));
+                        else
+                            LogHelper.LogWarning(LogHelper.FormatInvariant(LogMessages.IDX20806, (_metadataAddress ?? "null"), ex));
```
Can we bump this up to error?
How about LogHelper.LogException with the same message, but don't throw?

henrik-me replied:
Suggestion: Can this only happen if there is a network connection issue or how can it happen that we don't have the configuration? If we have additional information from e.g. the http response we could possibly add that? (e.g. time-out vs. 500, vs. not found, etc.)? Perhaps also depending on the response we can decide if it's a warning or an error?

---
In reply to: [280398515](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1157#discussion_r280398515) [](ancestors = 280398515)

brentschmaltz replied:
@henrik-me It depends on the retriever that was used. I have had conversations with users where they need details. For example, if an IdentityProvider accidentally pushes bad metadata {bad formatting for example}, that would cause an error.

mafurman replied:
@brentschmaltz It looks like the HttpDocumentRetriever that's used by default in the OpenIdConnectConfigurationRetriever ends up throwing if the response does not have a success status code. So that exception could have some extra information as to why retrieving metadata failed.

I'm not sure how we would go about providing additional information about the HTTP response from the ConfigurationManager, as this information would somehow need to be bubbled up from the IDocumentRetriever that's used by the configuration retriever...

henrik-me replied:
If that information at that point in time is logged then perhaps the logs already have enough information for the developer, even though it may not be used to potentially improve this error message. I think that is ok, however I suggest reviewing what we log in the HttpDocumentRetriever.

GeoK replied:
This discussion is closely related to https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1112.
Maria is correct, `EnsureSuccessStatusCode `method throws `HttpRequestException` that doesn't contain an http status code.

I believe that there is value in bubbling-up an exception that has an http status code property on it, which allows developers to react to specific events (e.g. 429 for throttling).

Tratcher replied:
https://github.com/dotnet/corefx/issues/26477. Please don't invent a new exception type, we can address this in the framework instead.

brentschmaltz replied:
@Tratcher As noted above, HttpDocumentRetriever calls EnsureSuccessStatusCode(), which throws. I have had customers indicate there is additional info as stated in https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1112. How do we serve that up?

mafurman replied:
I removed the new exception type and modified this PR to also address #1112.

henrik-me commented on src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager.cs:
```diff
@@ -199,7 +199,10 @@ public async Task<T> GetConfigurationAsync(CancellationToken cancel)
                         _syncAfter = DateTimeUtil.Add(now.UtcDateTime, _automaticRefreshInterval < _refreshInterval ? _automaticRefreshInterval : _refreshInterval);
-                        throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX20803, (_metadataAddress ?? "null")), ex));
+                        if (_currentConfiguration == null) // Throw an exception if there's no configuration to return.
+                            throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX20803, (_metadataAddress ?? "null")), ex));
```
>throw LogHelper. [](start = 28, length = 16)

Consider not throwing and logging here, this is happening further down with a similar message.

brentschmaltz replied:
I would like to continue to throw here as we could provide details that may not be available later.
I am suggesting we throw / inform users to understand metadata retrieval failed.

mafurman replied:
I was considering only throwing further down, but this will cause us to lose information about the original exception that was thrown.

henrik-me replied:
That information can be stored for later usage

henrik-me commented:
>                     throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX20803, (_metadataAddress ?? "null"))));

perhaps this should be the only place where _config == null should throw?

---
Refers to: src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager.cs:214 in b828af1. [](commit_id = b828af1b65f066f29d8c2694a715a2513290d489, deletion_comment = False)

mafurman commented:
I updated this PR so that a more specific exception type is thrown, and so that we log error instead of warning when we can't obtain an updated configuration (but the current configuration is not null).

I also added in a test case to check that the correct exception is thrown when we can't connect to the provided metadata endpoint. I think that the sequence of inner exceptions we provide does give enough information to understand why obtaining the configuration failed, but I can also work on providing more  information from the HTTP response in the exception thrown if we agree that's necessary.

Though I do believe that throwing a more specific exception (rather than InvalidOperationException) may be a better choice, I'm not sure if this is a change we want to make as it is a breaking change.

brentschmaltz commented:
:shipit:
